cmd/collect-profiles/main.go

```diff
@@ -0,0 +1,364 @@
+			}
+
+			newestConfigMaps, expiredConfigMaps := separateConfigMapsIntoNewestAndExpired(existingConfigMaps.Items)
+
+			// Attempt to delete all but the newest configMaps generated by this job
+			errs := []error{}
+			for _, cm := range expiredConfigMaps {
+				if err := cfg.Client.Delete(cmd.Context(), &cm); err != nil {
+					errs = append(errs, err)
+					continue
+				}
+				klog.Infof("Successfully deleted configMap %s/%s", cm.GetNamespace(), cm.GetName())
+			}
+
+			// If a delete call failed, abort to avoid creating new configMaps
+			if len(errs) != 0 {
+				return fmt.Errorf("error deleting expired pprof configMaps: %v", errs)
+			}
+
+			certPath := filepath.Join(certMountPath, corev1.TLSCertKey)
+			keyPath := filepath.Join(certMountPath, corev1.TLSPrivateKeyKey)
+
+			if err := verifyCertAndKeyExist(certPath, keyPath); err != nil {
+				klog.Infof("error verifying provided cert and key: %v", err)
+				klog.Info("generating a new cert and key")
+				return populateServingCert(cmd.Context(), cfg.Client)
+			}
+
+			httpClient, err := getHttpClient(certPath, keyPath)
+			if err != nil {
+				return err
+			}
+
+			// Track successfully created configMaps by generateName for each endpoint being scrapped.
+			createdCM := map[string]struct{}{}
+
+			for _, a := range validatedArguments {
+				b, err := requestURLBody(httpClient, a.url)
+				if err != nil {
+					klog.Infof("error retrieving pprof profile: %v", err)
+					continue
+				}
+
+				cm := &corev1.ConfigMap{
+					ObjectMeta: metav1.ObjectMeta{
+						GenerateName: a.generateName,
+						Namespace:    namespace,
+						Labels: map[string]string{
+							profileConfigMapLabelKey: "",
+						},
+					},
+					Immutable: getTruePointer(),
+					BinaryData: map[string][]byte{
+						"profile.pb.gz": b,
+					},
+				}
+
+				if err := cfg.Client.Create(cmd.Context(), cm); err != nil {
+					klog.Errorf("error created configMap %s/%s: %v", cm.GetNamespace(), cm.GetName(), err)
+					continue
+				}
+
+				klog.Infof("Successfully created configMap %s/%s", cm.GetNamespace(), cm.GetName())
+				createdCM[a.generateName] = struct{}{}
+			}
+
+			// Delete the configMaps which are no longer the newest
+			for _, cm := range newestConfigMaps {
+				// Don't delete ConfigMaps that were not replaced
+				// Also prevents deletes of configMaps with generateNames not included in command
+				if _, ok := createdCM[cm.GenerateName]; !ok {
+					continue
+				}
+				if err := cfg.Client.Delete(cmd.Context(), &cm); err != nil {
+					errs = append(errs, err)
+					continue
+				}
+				klog.Infof("Successfully deleted configMap %s/%s", cm.GetNamespace(), cm.GetName())
+			}
+
+			if len(errs) != 0 {
+				return fmt.Errorf("error deleting existing pprof configMaps: %v", errs)
+			}
+
+			// Update serving cert after a successful run
+			return populateServingCert(cmd.Context(), cfg.Client)
+		},
+	}
+}
+
+func verifyCertAndKeyExist(certPath, keyPath string) error {
+	fi, err := os.Stat(certPath)
```

Comment:
Any idea on whether os.Stat opens the file as well? It wasn't immediately clear to me what syscall was being used under-the-hood, and whether we need to explicitly close the file descriptor as well.

Reply:
Great question and something worth looking up! Based on the answer to [this Stack Overflow question](https://stackoverflow.com/questions/50311374/does-golangs-os-stat-open-the-file), os.Stat is a cross-platform abstraction for the Unix call stat which allows one to retrieve a metadata file's metadata without having to open it.

So no, one does not need to close the file.